File: packages/nodejs/src/endpoints/steambirdEndpoint.ts
import type { Endpoint } from "./endpoint.js";

export const SteambirdEndpoint = {
  notifications(
    issuer?: string,
    opts: { since?: string; until?: string } = {},
  ): Endpoint {
    const params: Record<string, string> = {};
    if (issuer != null) params["issuer"] = issuer;
    if (opts.since != null) params["since"] = opts.since;
    if (opts.until != null) params["until"] = opts.until;
    return {
      path: "/steambird/v1/notifications",
      method: "GET",
      queryParameters: Object.keys(params).length > 0 ? params : undefined,
    };
  },

  read(id: string): Endpoint {
    return { path: `/steambird/v1/notifications/${id}`, method: "POST" };
  },

  readAll(issuer?: string): Endpoint {
    return {
      path: "/steambird/v1/notifications/all",
      method: "POST",
      queryParameters: issuer != null ? { issuer } : undefined,
    };
  },

  unreads(issuer?: string, issuers: string[] = []): Endpoint {
    const params: Record<string, string> = {};
    if (issuer != null) params["issuer"] = issuer;
    if (issuers.length > 0) params["issuers"] = issuers.join(",");
    return {
      path: "/steambird/v1/notifications/unread",
      method: "GET",
      queryParameters: Object.keys(params).length > 0 ? params : undefined,
    };
  },
} as const;

File: packages/swift/Sources/CatalystSwift/Models/SteambirdEndpoint.swift
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki
import Foundation

public enum SteambirdEndpoint: Endpoint {
  case notifications(issuer: String?, since: String?, until: String?)
  case read(id: String)
  case readAll(issuer: String?)
  case unreads(issuer: String?, issuers: [String]?)

  public var path: String {
    switch self {
    case .notifications:
      return "/steambird/v1/notifications"

    case .read(let id):
      return "/steambird/v1/notifications/\(id)"

    case .readAll:
      return "/steambird/v1/notifications/all"

    case .unreads:
      return "/steambird/v1/notifications/unread"
    }
  }

  public var method: HTTPMethod {
    switch self {
    case .notifications, .unreads:
      return .get

    case .read, .readAll:
      return .post
    }
  }

  public var headers: [String: String]? {
    switch self {
    default:
      return nil
    }
  }

  public var queryParameters: [String: String]? {
    switch self {
    case .notifications(let issuer, let since, let until):
      var params: [String: String] = [:]
      if let issuer { params["issuer"] = issuer }
      if let since { params["since"] = since }
      if let until { params["until"] = until }
      return params.isEmpty ? nil : params

    case .readAll(let issuer):
      var params: [String: String] = [:]
      if let issuer { params["issuer"] = issuer }
      return params.isEmpty ? nil : params

    case .unreads(let issuer, let issuers):
      var params: [String: String] = [:]
      if let issuer { params["issuer"] = issuer }
      if let issuers { params["issuers"] = issuers.joined(separator: ",") }
      return params.isEmpty ? nil : params

    default:
      return nil
    }
  }

  public var body: (any Encodable)? {
    switch self {
    default:
      return nil
    }
  }
}
